File: frontend/lib/core/utils/validation_utils.dart
class ValidationUtils {
  /// Validates email based on user rules:
  /// - Optional (returns null if empty or null)
  /// - If provided, must end with '@gmail.com'
  static String? validateEmail(String? value) {
    if (value == null || value.trim().isEmpty) {
      return null; // Optional
    }
    final trimmed = value.trim();
    if (!trimmed.toLowerCase().endsWith('@gmail.com')) {
      return 'Email must end with @gmail.com';
    }
    return null;
  }

  /// Validates either email or phone for login
  static String? validateIdentifier(String? value) {
    if (value == null || value.trim().isEmpty) {
      return 'Email or Phone is required';
    }
    final trimmed = value.trim();
    if (trimmed.contains('@')) {
      return validateEmail(trimmed);
    } else {
      return validatePhone(trimmed);
    }
  }

  /// Validates phone number based on user rules:
  /// - Required
  /// - Must have exactly 9 digits after '0' or '+94'
  static String? validatePhone(String? value) {
    if (value == null || value.trim().isEmpty) {
      return 'Phone number is required';
    }
    final trimmed = value.trim();
    
    // Check if it starts with '0' and followed by 9 digits
    final zeroPattern = RegExp(r'^0[0-9]{9}$');
    // Check if it starts with '+94' and followed by 9 digits
    final intlPattern = RegExp(r'^\+94[0-9]{9}$');

    if (trimmed.startsWith('0')) {
      if (!zeroPattern.hasMatch(trimmed)) {
        return 'Phone number must have 9 digits after 0 (e.g. 0771234567)';
      }
    } else if (trimmed.startsWith('+94')) {
      if (!intlPattern.hasMatch(trimmed)) {
        return 'Phone number must have 9 digits after +94 (e.g. +94771234567)';
      }
    } else {
      return 'Phone number must start with 0 or +94';
    }
    
    return null;
  }

  /// Validates password based on user rules:
  /// - Required
  /// - At least 8 characters
  static String? validatePassword(String? value) {
    if (value == null || value.trim().isEmpty) {
      return 'Password is required';
    }
    if (value.trim().length < 8) {
      return 'Password must be at least 8 characters';
    }
    return null;
  }

  /// Validates required fields with a custom field name
  static String? validateRequired(String? value, String fieldName) {
    if (value == null || value.trim().isEmpty) {
      return '$fieldName is required';
    }
    return null;
  }

  static String? validatePrice(String? value) {
    if (value == null || value.trim().isEmpty) {
      return 'Price is required';
    }
    final price = double.tryParse(value);
    if (price == null || price <= 0) {
      return 'Enter a valid price greater than 0';
    }
    return null;
  }

  static String? validateStock(String? value) {
    if (value == null || value.trim().isEmpty) {
      return 'Stock level is required';
    }
    final stock = int.tryParse(value);
    if (stock == null || stock < 0) {
      return 'Enter a valid non-negative number';
    }
    return null;
  }
}

File: backend/src/utils/validationUtils.js
/**
 * Backend validation utilities for auth.
 */

const isValidEmail = (email) => {
    if (!email) return true; // Optional
    return email.toLowerCase().endsWith('@gmail.com');
};

const isValidPhone = (phone) => {
    if (!phone) return false; // Required
    const trimmed = phone.trim();
    // 0771234567 (10 digits) or +94771234567 (12 chars)
    return /^0[0-9]{9}$/.test(trimmed) || /^\+94[0-9]{9}$/.test(trimmed);
};

const isValidPassword = (password) => {
    return password && password.length >= 8;
};

// Ensures the provided price is a positive numeric value for products/sales.
const isValidPrice = (price) => {
    const val = parseFloat(price);
    // Use isNaN check and ensure it's strictly greater than zero.
    return !isNaN(val) && val > 0;
};

// Ensures stock levels are valid non-negative integers.
const isValidStock = (stock) => {
    // Initial null/undefined check for required inventory fields.
    if (stock === null || stock === undefined || stock === '') return false;
    const val = Number(stock);
    // Integers are required for discrete units (e.g. 1 item, 2 items).
    return Number.isInteger(val) && val >= 0;
};

module.exports = {
    isValidEmail,
    isValidPhone,
    isValidPassword,
    isValidPrice,
    isValidStock
};
